xds/src/main/java/com/linecorp/armeria/xds/ClusterResourceHolder.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.envoyproxy.envoy.config.cluster.v3.Cluster;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext;
+
+/**
+ * A cluster object for a {@link Cluster}.
+ */
+public final class ClusterResourceHolder extends AbstractResourceHolder {
+
+    private final Cluster cluster;
+    @Nullable
+    private final RouteResourceHolder primer;
+
+    ClusterResourceHolder(Cluster cluster) {
+        this(cluster, null);
+    }
+
+    ClusterResourceHolder(Cluster cluster, @Nullable RouteResourceHolder primer) {
+        this.cluster = cluster;
+        this.primer = primer;
+    }
+
+    @Override
+    public ClusterResourceHolder withPrimer(@Nullable ResourceHolder primer) {
+        if (primer == null) {
+            return this;
+        }
+        checkArgument(primer instanceof RouteResourceHolder);
+        return new ClusterResourceHolder(cluster, (RouteResourceHolder) primer);
+    }
+
+    @Nullable
+    UpstreamTlsContext upstreamTlsContext() {
+        if (cluster.hasTransportSocket()) {
+            final String transportSocketName = cluster.getTransportSocket().getName();
+            checkArgument("envoy.transport_sockets.tls".equals(transportSocketName),
+                          "Unexpected tls transport socket name '%s'", transportSocketName);
+            try {
+                return cluster.getTransportSocket().getTypedConfig()
+                              .unpack(UpstreamTlsContext.class);
+            } catch (Exception e) {
+                throw new RuntimeException("Error unpacking tls context", e);
```

Comment:
How about using `IllegalStateException` instead?

Reply:
Used `IllegalArgumentException` instead to unify behavior with other validation methods